atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RangeLoader.java

```diff
@@ -65,20 +69,36 @@ public void setConsistencyLevel(ConsistencyLevel consistencyLevel) {
             ConsistencyLevel consistency,
             Supplier<ResultsExtractor<T>> resultsExtractor) {
         SlicePredicate predicate;
+
         if (rangeRequest.getColumnNames().size() == 1) {
-            byte[] colName = rangeRequest.getColumnNames().iterator().next();
-            predicate = SlicePredicates.latestVersionForColumn(colName, startTs);
+            predicate = getSlicePredicateForSingleColumn(rangeRequest, startTs);
         } else {
-            // TODO(nziebart): optimize fetching multiple columns by performing a parallel range request for
-            // each column. note that if no columns are specified, it's a special case that means all columns
-            predicate = SlicePredicates.create(SlicePredicates.Range.ALL, SlicePredicates.Limit.NO_LIMIT);
+            TableMetadata tableMetadata = metaDataProvider.getMetaDataFromRef(tableRef);
```

Comment:
It's a bit shakey to rely on table metadata here. Basically, this has never been relied upon for correctness of the sort you're trying to use it for, and so in classic Atlas fashion it will almost surely break somewhere (I can give you an internal example where I know it will).

More reliable is using the code generation, because it's at an API layer which will ignore any additional columns returned. This is the PR that probably missed it the first time around.

https://github.com/palantir/atlasdb/pull/713/files

Reply:
or, to copy here my internal comment

```
    /**
     * {@link NameMetadataDescription} and {@link ColumnMetadataDescription} are only used by atlasdb-cli to interact
     * with the internal representation of tables. They do not specify actual shape of tables.
     */
```

and so you probably don't get to assume that the information is correct :P

Reply:
because of #713 do we expect this change (the whole PR) to actually change performance for single column tables? as in is the codepath we're optimizing here even possible to hit? the generated code should sense that for any "all columns" request that it should specify the specific set of columns

Reply:
In general, different nodes can end up diverging on metadata information, since they read it once on startup and then never again (unless that's changed). There can be fun issues with `noColumns()` tables here as well...